### DISCUSSION TITLE
Extent of Feature

### DISCUSSION THREAD

**Discussion by charlenni** [Mapsui/Mapsui#2450] [Development]
It would be nice to have the extent of a feature. This is helpful in many places.
1. It is possible to check, very fast, if a touch hit a feature or not without drawing it. So only features, that are possible to hit are drawn.
2. It is also possible to draw features that are outside the viewport, if their extent is inside. For this, we use a default symbol size of 64 px.

Is this a good idea? Any suggestions?

**Comment by pauldendulk** [comment]
There is an IFeature.Extent, but this is based on the geometry, and the width is zero for points. I guess you are referring to the extent of the feature the way it is drawn. Taking into account the symbol, the scale of the symbol, the width of the of the line of a polygon or polyline. Perhaps even line ending. 

This may not be easy to determine. For practical purposes it does not need to perfectly fit the geometry, it could be bigger (but not smaller). For linewidth of polygons and polylines it is possible to work with a big margin, just to be sure. It won't make a big difference for performance. Guess, the most important use case is for symbols. 

We could calculate that extent in map coordinates. But a problem with that is that it is different depending on the zoom resolution. Perhaps we just want to know the extent in pixels, and I think relative to the symbol origin, so you need to calculate it only once. What were you thinking about?

The default symbol size of 64 is definitely awkward and has always been bothering me.

**Comment by charlenni** [comment]
I think about symbols. There it should be possible without using the renderer. To be clear: it should work with all features that didn't depend on the map zoom factor.

We have two types of features:
1. Point
2. All others

Point features are normally symbols, which don't change their size when changing the zoom factor. There we can easily calculate the extent. The other type changes their size with the zoom factor and could be tested like we already do.

When I think about this, then I have the behavior of vector maps in mind. They have many symbols (with and without text) on the screen. They have rules when to display and when to not display them or parts of them. Furthermore, they have a hierarchy of importance (more important remove lesser important ones).

**Reply by pauldendulk** [reply]
ah right. That is a good distinction. You basically want the proper way to determine the 64x. 

Currently the 64px is important for fetching symbols outside the extent that are partially visible within the extent. The current solution grows the extent based on the 64px. The problem here is that you need to know the extent of items that you did not fetch yet. You do not know they exist. An alternative would be to do the query with the normal extent and let the DataSource take into account the symbol size. I don't fully see how that will work. Most of the time we want to query a database, which contains just the point. You could add symbol information to the database, but the symbol can be dynamic. They can have different sizes depending on the state. In the apps that I use symbols can contain various parts. If the status of a vehicle is 'Loading' for 2 minutes the symbol will be different during that period.
